### ISSUE TITLE
New .admonition/.details injects padding into *any* last-child

### ISSUE THREAD

**Issue by facelessuser** [squidfunk/mkdocs-material#630]
## Description

Currently Material is assigning any element that is a *last-child* with padding.  This is bad.  So the math elements have some children who or *last-child* to their parent even if they are not direct children. of `.admonition`.  Since they are descendants, they inherit the padding.

Something like `.admonition>:last-child` and `details>:last-child` would be more appropriate. as it would limit the padding the last *direct* child.

Please see issue: https://github.com/facelessuser/pymdown-extensions/issues/187.

### Expected behavior

Padding should *only* be applied to the last direct descendant of `.admonition` or `details`.

### Actual behavior

Any element that qualifies as a *last-child* is inheriting bottom padding.

### Steps to reproduce the bug

1. NA

### Package versions

* Material: 2.2.1 &amp; 2.2.2

### Project configuration

https://github.com/facelessuser/pymdown-extensions/blob/master/mkdocs.yml

### System information

* OS: NA
* Browser: NA

**Comment by facelessuser** [comment]
On a side note, I also question this case as it is also very aggressive, but I don't think I see any side effects of it, but maybe should be evaluated.

```css
.md-typeset .admonition :first-child, .md-typeset details :first-child {
    margin-top: 0;
}
```

**Comment by facelessuser** [comment]
I'll probably try and throw together a pull this week.  I figure the `first-child` should probably be `parent>:nth-child(2)` and `last-child` should probably be `parent>:last-child`.

**Comment by makshh** [comment]
```css
.md-typeset .admonition :first-child,
.md-typeset details :first-child {
  margin-top: 0;
}
```

Is this necessary at all?

**Comment by facelessuser** [comment]
@makshh, I assume it is there to limit some elements that add extra margin to their top initially.  I suspect the idea is the first block element should not have extra padding in an admonition or details element to keep a consistent look.

I practicality, I'm not sure if there are specific elements in Material that are adding extra margins that need to be stripped.  I assume it was added for a reason, but I don't know.

**Comment by makshh** [comment]
There is also a problem with specificity, this screenshot is from official GitHub page.

![image](https://user-images.githubusercontent.com/5025950/33800921-5d9f9b1e-dd4b-11e7-924d-547551bf34e0.png)

And regarding to your fixes, I think `:nth-child(2)` won't be good.
Example: admonition without title with two paragraphs (second paragraph will have `margin-top` of 0 unless it has higher specificity).

**Comment by facelessuser** [comment]
Hmm, that leads me to believe that it is targeting the title element.  If that is its real purpose, then that could be targeted instead with `.admonition>.admonition-title` and `details>summary`  I know that details is kind of inheriting style from admonitions, so I wonder it was made more general to suit the both.

**Comment by squidfunk** [comment]
Attempted a fix in PR #634, specifically in [this file](https://github.com/squidfunk/mkdocs-material/pull/634/files#diff-475e4c5b939319f09d50d708949a03ef). The margin issue @makshh mentioned is also fixed. The problem was that, indeed, all margin was applied to all nested last childs. This should be fixed by the child selector. Just merged the PR into master - could you validate?

**Comment by facelessuser** [comment]
Fantastic!  Meant to get to this myself, but you beat me to it 🙂.

**Comment by squidfunk** [comment]
Did you test it? Can I make a release with a fix? I lack further use cases :-)

**Comment by facelessuser** [comment]
I will formally test your changes and let you know, but I had essentially tested both of your changes directly in the browser when I was investigating.

**Comment by makshh** [comment]
I think it is fixed.

Possible cases when it may not work as expected:

1. Big top margin on first child (for example if you include heading inside admonition it will have big top margin).

2. Last child of the last element has margin top (so the margin will be margin of that element + 1.2rem defined in the fix).

But I think it's acceptable as it is now.

**Comment by facelessuser** [comment]
@squidfunk, my use cases are fixed. 👍

**Comment by squidfunk** [comment]
Thanks for your feedback! Released as part of 2.2.3.
